src/hal/src/pso/descriptor.rs

```diff
@@ -95,33 +93,31 @@ pub trait DescriptorPool<B: Backend>: Send + Sync + fmt::Debug {
         I: IntoIterator,
         I::Item: Borrow<B::DescriptorSetLayout>,
     {
-        layouts.into_iter().map(|layout| self.allocate_set(layout.borrow())).collect()
+        layouts
+            .into_iter()
+            .map(|layout| self.allocate_set(layout.borrow()))
+            .collect()
     }
 
     ///
     fn reset(&mut self);
 }
 
 #[allow(missing_docs)] //TODO
-pub struct DescriptorSetWrite<'a, B: Backend, R: 'a + RangeArg<u64>> {
+pub struct DescriptorSetWrite<'a, B: Backend, W> {
     pub set: &'a B::DescriptorSet,
     pub binding: usize,
-    pub array_offset: usize,
-    pub write: DescriptorWrite<'a, B, R>,
+    pub writes: W,
 }
 
 #[allow(missing_docs)] //TODO
-#[derive(Clone, Copy)]
-pub enum DescriptorWrite<'a, B: Backend, R: 'a + RangeArg<u64>> {
-    Sampler(&'a [&'a B::Sampler]),
-    SampledImage(&'a [(&'a B::ImageView, ImageLayout)]),
-    StorageImage(&'a [(&'a B::ImageView, ImageLayout)]),
-    InputAttachment(&'a [(&'a B::ImageView, ImageLayout)]),
-    UniformBuffer(&'a [(&'a B::Buffer, R)]),
-    StorageBuffer(&'a [(&'a B::Buffer, R)]),
-    UniformTexelBuffer(&'a [&'a B::BufferView]),
-    StorageTexelBuffer(&'a [&'a B::BufferView]),
-    CombinedImageSampler(&'a [(&'a B::Sampler, &'a B::ImageView, ImageLayout)]),
+#[derive(Clone)]
+pub enum Descriptor<'a, B: Backend> {
+    Sampler(&'a B::Sampler),
+    Image(&'a B::ImageView, ImageLayout),
+    CombinedImageSampler(&'a B::ImageView, ImageLayout, &'a B::Sampler),
+    Buffer(&'a B::Buffer, Range<Offset>),
```

Comment:
`Range<Offset>` can be replaced by `(Option<Offset>, Option<Offset>)`  in which any range-like type can be transformed by some convenient method of `RangeArg` trait.